### PR TITLE
GDPR-112 Correcting pessimistic lock and proving behaviour with integration test

### PR DESCRIPTION
**Pull request by brightonsbox**
I found that the previous method of using the EntityManager to implement a pessimistic lock was failing with `java.lang.IllegalArgumentException: entity not in the persistence context`.

Essentially because the entity was not originally retrieved by the EntityManager, it was outside its 'persistence context'.  This PR gets around this issue and implements an integration test to demonstrate the behaviour.

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/referral/ReferralResolutionRepository.java [hunk 1]
```diff
@@ -5,5 +5,5 @@
 import uk.gov.justice.hmpps.datacompliance.repository.jpa.model.referral.ReferralResolution;
 
 @Repository
-public interface ReferralResolutionRepository extends CrudRepository<ReferralResolution, Long> {
+public interface ReferralResolutionRepository extends CrudRepository<ReferralResolution, Long>, ReferralResolutionRepositoryCustom {
```
Comment (brightonsbox) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/repository/jpa/repository/referral/ReferralResolutionRepository.java:8]: This follows the repository customisation approach here: https://docs.spring.io/spring-data/data-commons/docs/2.4.0-SNAPSHOT/reference/html/#repositories.custom-implementations allowing access to the EntityManager lock method.

File: src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionService.java [hunk 2]
```diff
@@ -64,11 +62,12 @@ public void processReferral(final OffenderDeletionReferral referral,
 
     public void processUpdatedRetentionCheck(final RetentionCheck retentionCheck) {
 
-        final var referralResolution = retentionCheck.getReferralResolution();
+        final var referralResolution =
+                referralResolutionRepository.findById(retentionCheck.getReferralResolution().getResolutionId())
+                        .orElseThrow();
```
Comment (brightonsbox) [inline src/main/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionService.java:67]: By finding the 'resolution' entity using the id, it is now within the persistence context and we can lock the entity.

File: src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionServiceIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    void processUpdatedRetentionCheckResultsInRetention() {
+
+        assertThat(referralResolutionRepository.findById(1L).orElseThrow().getResolutionStatus()).isEqualTo(PENDING);
+
+        CompletableFuture.allOf(
+                runAsync(() -> updateAndProcessCheck(1L)),
+                runAsync(() -> updateAndProcessCheck(2L)))
+                .join();
+
+        assertThat(referralResolutionRepository.findById(1L).orElseThrow().getResolutionStatus()).isEqualTo(RETAINED);
```
Comment (brightonsbox) [inline src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionServiceIntegrationTest.java:44]: If the pessimistic lock isn't used, a race condition can mean that the status remains as PENDING.  I've proved this is the case by (locally) putting some carefully timed `Thread.sleep` s in so it always happens.  By introducing the lock though, this issue is resolved.

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -74,4 +74,4 @@ management:
       cache:
         time-to-live: 2000ms
 
-graceful.shutdown.enabled: true
+server.shutdown.grace-period: 10s
```
Comment (mikehalmamoj) [inline src/main/resources/application.yml:77]: I'm guessing you copied this from my PR? Sorry it's wrong! https://docs.spring.io/spring-boot/docs/2.3.0.RELEASE/reference/htmlsingle/#boot-features-graceful-shutdown

File: src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionServiceIntegrationTest.java (hunk 3 above)
Comment (mikehalmamoj) [reply src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionServiceIntegrationTest.java:44]: I realise this is not easy... but as this test doesn't always fail, is there any way to make it?  For instance by using a SpyBean to reproduce the sleeps that you've proved it with locally? Just worried that someone could remove the lock and not realise (though at least you have fixed the bug - and not worth spending loads of time on this test).

Comment (brightonsbox) [reply src/test/java/uk/gov/justice/hmpps/datacompliance/services/referral/ReferralResolutionServiceIntegrationTest.java:49]: I have managed to ensure that the test will fail if the lock removed, good suggestion 👍

**Pull request merged**